File: fastlimit/algorithms/token_bucket.py
"""
Token Bucket rate limiting algorithm implementation.

The token bucket algorithm provides smoother rate limiting compared to
fixed window, with better handling of bursty traffic.
"""

import logging
import time
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..backends.redis import RedisBackend

from .base import RateLimitAlgorithm, RateLimitResult

logger = logging.getLogger(__name__)


class TokenBucket(RateLimitAlgorithm):
    """
    Token Bucket algorithm implementation.

    How it works:
    - A bucket holds tokens up to a maximum capacity
    - Tokens are continuously added at a fixed refill rate
    - Each request consumes one or more tokens
    - If not enough tokens, request is denied
    - Provides smooth rate limiting without boundary bursts

    Example:
        100 requests/minute = 100 max tokens, ~1.67 refill rate/second
        Allows bursts up to 100 requests, then sustained 1.67 req/sec

    Advantages over Fixed Window:
    - No burst at window boundaries
    - Smoother traffic distribution
    - Better for bursty workloads
    - Allows controlled bursts within capacity

    Disadvantages:
    - Slightly more complex
    - Uses more Redis memory (stores tokens + timestamp)
    - Can allow more requests in first window
    """

    def __init__(self, backend: "RedisBackend") -> None:
        """
        Initialize Token Bucket algorithm.

        Args:
            backend: Redis backend for executing Lua scripts
        """
        self.backend = backend
        logger.debug("Initialized TokenBucket algorithm")

    async def check(
        self,
        key: str,
        max_requests: int,
        window_seconds: int,
        cost: int = 1000,
    ) -> RateLimitResult:
        """
        Check if a request is allowed under token bucket rate limit.

        Args:
            key: Unique identifier for the rate limit
            max_requests: Maximum tokens (bucket capacity, with multiplier)
            window_seconds: Time window to spread requests over
            cost: Number of tokens to consume (with multiplier)

        Returns:
            RateLimitResult with allowed status and metadata

        Example:
            For "100/minute":
            - max_requests = 100000 (100 * 1000)
            - window_seconds = 60
            - refill_rate = 100000 / 60 = 1666.67 tokens/sec
            - Bucket starts full (100000 tokens)
            - Refills at 1666.67 tokens/sec
            - Max capacity: 100000 tokens
        """
        # Calculate refill rate (tokens per second as integer)
        # For 100/minute: 100000 / 60 = 1666 tokens/sec
        refill_rate_per_second = max_requests // window_seconds

        # Get current timestamp in milliseconds
        current_time_ms = int(time.time() * 1000)

        # Execute token bucket Lua script
        result = await self.backend.check_token_bucket(
            key=key,
            max_tokens=max_requests,
            refill_rate_per_second=refill_rate_per_second,
            window_seconds=window_seconds,
            current_time_ms=current_time_ms,
            cost=cost,
        )

        # Calculate reset timestamp (when bucket would be full)
        # If tokens remaining, no reset needed
        # If denied, reset_at = current_time + retry_after
        reset_at = None
        if not result.allowed:
            reset_at = (current_time_ms // 1000) + (result.retry_after // 1000)

        return RateLimitResult(
            allowed=result.allowed,
            remaining=result.remaining,
            retry_after=result.retry_after,
            reset_at=reset_at,
        )

    async def reset(self, key: str) -> bool:
        """
        Reset token bucket for a specific key.

        This removes the bucket data, causing the next request
        to start with a full bucket.

        Args:
            key: Unique identifier for the rate limit

        Returns:
            True if reset was successful
        """
        return await self.backend.reset(key)

    async def get_usage(self, key: str, *args: Any, **kwargs: Any) -> dict[str, Any]:
        """
        Get current usage statistics for a token bucket.

        Args:
            key: Unique identifier for the rate limit
            *args: Expected to contain max_requests as first argument
            **kwargs: Not used

        Returns:
            Dictionary with:
            - current: Current token count (without multiplier)
            - limit: Maximum tokens (without multiplier)
            - remaining: Tokens remaining (without multiplier)
            - last_refill: Unix timestamp of last refill

        Note: For token bucket, "current" means tokens available,
        not requests consumed (inverse of fixed window).
        """
        max_requests: int = args[0] if args else kwargs.get("max_requests", 0)
        usage = await self.backend.get_token_bucket_usage(key)

        # Convert from integer math (divide by 1000)
        current_tokens = usage.get("tokens", max_requests) // 1000
        max_tokens_display = max_requests // 1000

        return {
            "current": current_tokens,
            "limit": max_tokens_display,
            "remaining": current_tokens,
            "last_refill": usage.get("last_refill", 0),
        }


def calculate_refill_rate(requests: int, window_seconds: int) -> float:
    """
    Calculate token refill rate for token bucket.

    Args:
        requests: Number of requests allowed in window
        window_seconds: Time window in seconds

    Returns:
        Refill rate in tokens per second

    Examples:
        >>> calculate_refill_rate(100, 60)  # 100/minute
        1.6666666666666667

        >>> calculate_refill_rate(1000, 3600)  # 1000/hour
        0.2777777777777778

        >>> calculate_refill_rate(10, 1)  # 10/second
        10.0
    """
    if window_seconds <= 0:
        raise ValueError("Window seconds must be positive")

    return requests / window_seconds


def calculate_bucket_capacity(requests: int, burst_factor: float = 1.0) -> int:
    """
    Calculate bucket capacity with optional burst allowance.

    By default, bucket capacity equals the rate limit. You can
    increase capacity to allow larger bursts.

    Args:
        requests: Base number of requests allowed
        burst_factor: Multiplier for bucket capacity (>= 1.0)
                     1.0 = no extra burst (default)
                     2.0 = allow 2x burst
                     0.5 = not recommended (less than rate)

    Returns:
        Bucket capacity in tokens

    Examples:
        >>> calculate_bucket_capacity(100, 1.0)  # Standard
        100

        >>> calculate_bucket_capacity(100, 1.5)  # Allow 50% burst
        150

        >>> calculate_bucket_capacity(100, 2.0)  # Allow 2x burst
        200

    Note: Burst factor is not currently exposed in the main API,
    but is available for advanced use cases.
    """
    if burst_factor < 1.0:
        logger.warning(
            f"Burst factor {burst_factor} < 1.0 may cause issues. " "Consider using >= 1.0"
        )

    return int(requests * burst_factor)

File: fastlimit/scripts/token_bucket.lua
-- Token Bucket Rate Limiting Script
-- Implements token bucket algorithm with atomic operations
--
-- KEYS[1] = rate limit key (e.g., "ratelimit:tenant123:premium:bucket")
-- ARGV[1] = max_tokens (bucket capacity, e.g., 100000 for 100 tokens with 1000x multiplier)
-- ARGV[2] = refill_rate_per_second (tokens per second, integer with 1000x multiplier)
-- ARGV[3] = window_seconds (window duration for TTL calculation)
-- ARGV[4] = current_time_ms (current timestamp in milliseconds)
-- ARGV[5] = cost (tokens to consume, e.g., 1000 for cost=1 with 1000x multiplier)
--
-- Returns: {allowed (1 or 0), remaining, retry_after_ms}
--
-- Token Bucket Algorithm:
-- - Tokens are continuously added at refill_rate
-- - Bucket has maximum capacity of max_tokens
-- - Each request consumes 'cost' tokens
-- - If not enough tokens, request is denied
-- - Uses millisecond precision to support low rates (e.g., 1/hour)

local key = KEYS[1]
local max_tokens = tonumber(ARGV[1])
local refill_rate_per_second = tonumber(ARGV[2])
local window_seconds = tonumber(ARGV[3])
local current_time_ms = tonumber(ARGV[4])
local cost = tonumber(ARGV[5]) or 1000  -- Default to 1000 (cost=1) if not provided

-- Get current bucket state
local bucket = redis.call('HMGET', key, 'tokens', 'last_refill_ms')
local current_tokens = tonumber(bucket[1]) or max_tokens  -- Start with full bucket
local last_refill_ms = tonumber(bucket[2]) or current_time_ms

-- Calculate time elapsed since last refill (in milliseconds)
local time_elapsed_ms = math.max(0, current_time_ms - last_refill_ms)

-- Calculate tokens to add based on elapsed time
-- tokens_to_add = refill_rate_per_second * (time_elapsed_ms / 1000)
-- Use integer math: tokens_to_add = (refill_rate_per_second * time_elapsed_ms) / 1000
local tokens_to_add = 0
if refill_rate_per_second > 0 then
    tokens_to_add = math.floor((refill_rate_per_second * time_elapsed_ms) / 1000)
end

-- Add tokens to bucket, but don't exceed max capacity
local new_tokens = math.min(max_tokens, current_tokens + tokens_to_add)

-- Determine if request is allowed
local allowed = 0
local remaining = 0
local retry_after_ms = 0

if new_tokens >= cost then
    -- Request is allowed - consume tokens
    allowed = 1
    new_tokens = new_tokens - cost
    remaining = new_tokens

    -- Update bucket state with millisecond timestamp
    redis.call('HMSET', key, 'tokens', new_tokens, 'last_refill_ms', current_time_ms)

    -- Set expiry to prevent memory leaks
    -- Use window_seconds * 2 as a safe TTL (bucket expires after inactivity)
    local ttl = window_seconds * 2 + 60  -- Add 60s buffer
    redis.call('EXPIRE', key, ttl)
else
    -- Request is denied - not enough tokens
    allowed = 0
    remaining = 0

    -- Calculate how long until enough tokens are available
    local tokens_needed = cost - new_tokens
    if refill_rate_per_second > 0 then
        -- time_needed_ms = (tokens_needed / refill_rate_per_second) * 1000
        -- Integer math: time_needed_ms = (tokens_needed * 1000) / refill_rate_per_second
        retry_after_ms = math.ceil((tokens_needed * 1000) / refill_rate_per_second)
    else
        -- If refill rate is 0 (shouldn't happen), wait for full window
        retry_after_ms = window_seconds * 1000
    end

    -- Update bucket state with refilled tokens (even though request denied)
    redis.call('HMSET', key, 'tokens', new_tokens, 'last_refill_ms', current_time_ms)

    -- Set expiry
    local ttl = window_seconds * 2 + 60
    redis.call('EXPIRE', key, ttl)
end

-- Return results
-- allowed: 1 if request should proceed, 0 if rate limited
-- remaining: number of tokens remaining in bucket (with multiplier)
-- retry_after_ms: milliseconds until enough tokens available (0 if allowed)
return {allowed, remaining, retry_after_ms}
